src/server/master_server/master_chunk_server_manager_service_impl.cc

```diff
@@ -10,13 +15,126 @@ using protos::grpc::ReportChunkServerRequest;
     ServerContext* context, const ReportChunkServerRequest* request,
     ReportChunkServerReply* reply) {
-  // TODO(bmokutub): handle the request by using the ChunkServerManager for
-  // updating chunk/chunkserver map and metadata manager to check chunk latest
-  // version.
+  auto& new_server_info = request->chunk_server();
+  auto existing_server_info =
+      gfs::server::ChunkServerManager::GetInstance().GetChunkServer(
+          new_server_info.location());
+
+  if (!existing_server_info.first) {
+    // Not found, new server info, maybe chunkserver is just starting up or
+    // restarted. Or the master just started or crashed/restarted and lost all
+    // in-mem chunkservermgr record. Lets register this chunkserver.
+    std::shared_ptr<protos::ChunkServer> new_chunk_server(
+        new protos::ChunkServer(new_server_info));
+
+    gfs::server::ChunkServerManager::GetInstance().RegisterChunkServer(
+        new_chunk_server);
+
+  } else {
+    // The chunkserver already exist, so we need to compare the existing info
+    // with new info and update.
+
+    auto existing_server = existing_server_info.first;
+    auto existing_server_lock = existing_server_info.second;
+
+    // The reported chunks that we don't know about.
+    absl::flat_hash_set<std::string> chunks_to_add;
+
+    // The chunks that we think exist on the chunkserver but no longer exist
+    // on it.
+    absl::flat_hash_set<std::string> chunks_to_remove;
+
+    // Lock guard scope. We need to drop this read lock before calling update
+    // chunkserver.
+    {
+      // First set all the reported chunks as chunks to add to our chunkserver
+      // info.
+      chunks_to_add = CreateFlatSetFromRepeatedPtrField<std::string>(
+          request->chunk_server().stored_chunk_handles());
+
+      // We need to take a read lock because chunkservermgr could be doing
+      // allocation for this server at the same time we are trying to read.
+      absl::ReaderMutexLock chunk_server_read_lock_guard(
+          existing_server_lock.get());
+
+      // Compare with our stored chunk handles for the chunk server. To see
+      // which reported chunks we have or don't have.
+      for (int i = 0; i < existing_server->stored_chunk_handles_size(); ++i) {
+        auto current_chunk_handle = existing_server->stored_chunk_handles(i);
+
+        if (chunks_to_add.contains(current_chunk_handle)) {
+          // Chunk server also reported this chunk. We know about this chunk,
+          // no need to add. We remove it from the add set. We will know what
+          // reported chunks weren't found in chunkservermgr at the end of the
+          // iteration.
+          // TODO(bmokutub): Check with metadata mgr that reported version
+          // matches the current version if it doesn't match, add to stale
+          // chunks in reply so chunkserver can delete it. And remove this
+          // chunkserver from the chunk locations in chunkservermgr, so we don't
```

Comment:
also, if chunk server has a newer version than the master, it means master crashed just after asking all chunk servers to advance chunk version before a write, so the master in this case simply advance his own version number

Reply:
yeah true. Will include that in the comment.